### ISSUE TITLE
Eigenvalues reported by circulant embedding algorithm

### ISSUE THREAD

**Issue by devmotion** [PieterjanRobbe/GaussianRandomFields.jl#11]
Is it intended that the circulant embedding algorithm does not report the actual eigenvalues (or rather the minimal eigenvalue) but a scaled value of it?

I implemented a circulant embedding algorithm and noticed that the eigenvalues reported by GaussianRandomFields did not match my results (not even closely when using a large number of grid points). The problem is that the inverse Fourier transform `irfft` in https://github.com/PieterjanRobbe/GaussianRandomFields.jl/blob/master/src/generators/circulant_embedding.jl#L72 is scaled by the number of elements (see https://juliamath.github.io/AbstractFFTs.jl/stable/api.html#AbstractFFTs.ifft) whereas the eigenvalues of the circulant matrix can be computed by applying an unnormalized multi-dimensional (inverse) Fourier transform to the first column of the circulant matrix, as e.g. discussed in great detail by Graham et al. in "Quasi-Monte Carlo methods for elliptic PDEs with random coefficients and applications" (2011). A small example in one dimension:
```julia
julia> using ToeplitzMatrices, FFTW, LinearAlgebra

julia> c = [1,2,3,4,3,2];

julia> fft(c)
6-element Array{Complex{Float64},1}:
 15.0 + 0.0im
 -4.0 + 0.0im
  0.0 + 0.0im
 -1.0 + 0.0im
  0.0 + 0.0im
 -4.0 + 0.0im

julia> ifft(c)
6-element Array{Complex{Float64},1}:
                  2.5 + 0.0im
  -0.6666666666666666 + 0.0im
                  0.0 + 0.0im
 -0.16666666666666666 + 0.0im
                  0.0 + 0.0im
  -0.6666666666666666 + 0.0im

julia> bfft(c)
6-element Array{Complex{Float64},1}:
 15.0 + 0.0im
 -4.0 + 0.0im
  0.0 + 0.0im
 -1.0 + 0.0im
  0.0 + 0.0im
 -4.0 + 0.0im

julia> eigvals(Matrix(Circulant(c)))
6-element Array{Float32,1}:
 -3.9999998  
 -3.9999988  
 -1.0000002  
 -4.644673e-8
  2.848653e-7
 15.000002   
```

As far as I can see, this error does not invalidate the generated random fields since the "eigenvalues" are not scaled in https://github.com/PieterjanRobbe/GaussianRandomFields.jl/blob/master/src/generators/circulant_embedding.jl#L86 during setup or when generating the random fields, which would be required otherwise.

However, since the size of the eigenvalues might affect how much padding is applied the reported eigenvalue should be fixed in my opinion, either by using `brfft` instead of `irfft` and applying the scaling in https://github.com/PieterjanRobbe/GaussianRandomFields.jl/blob/master/src/generators/circulant_embedding.jl#L86 or by scaling the reported eigenvalue.

Unrelated to this problem, I noticed that the current implementation does not follow the algorithm presented by Graham et al. completely since the mirroring introduces a duplicated boundary in two and more dimensions (the 2n+1 grid points in dimensions 2,...,d could be reduced to 2n grid points, which would also simplify https://github.com/PieterjanRobbe/GaussianRandomFields.jl/blob/master/src/generators/circulant_embedding.jl#L68-L71 since the embedding in `c̃` would not be required anymore). Moreover, it seems one can get rid of mirroring altogether by exploiting the symmetry of c and using `FFTW.r2r(c, FFTW.REDFT00)` instead of `irfft`.

**Comment by PieterjanRobbe** [comment]
Hi David

Thanks for these interesting comments. Indeed, the current implementation does not follow the algorithm from the paper exactly. I don't know what the effect is of the scaling on the amount of padding, but I guess it doesn't hurt changing `irfft` into `brfft`.

Regarding your second comment, I think this is only possible in case of an isotropic random field? In that case, the covariance function only depends on the (2-norm of the) distance between the points, and only the first column of the covariance matrix is required. I should check my notes again to see why this is not so for general random fields. Correct me if I'm wrong here!

**Comment by devmotion** [comment]
>  I don't know what the effect is of the scaling on the amount of padding, but I guess it doesn't hurt changing irfft into brfft.

I just want to stress that for a 50x50 grid with the same covariance function and without padding GaussianRandomFields and my custom implementation (in which I tried to implement the algorithm from the paper exactly) report minimal eigenvalues that differ by a factor of approximately (2 * 50)^2, i.e. 1e4 (for a concrete example with a squared exponential covariance function GaussianRandomFields reports around -0.017 whereas I obtain -140). So my concern was that changing an "eigenvalue" of -0.017 to 0 seems to affect the statistical properties of the samples a lot less than changing an eigenvalue of -140 to 0.

> Regarding your second comment, I think this is only possible in case of an isotropic random field? In that case, the covariance function only depends on the (2-norm of the) distance between the points, and only the first column of the covariance matrix is required. I should check my notes again to see why this is not so for general random fields. Correct me if I'm wrong here!

Isotropy is not needed but this simplification requires that the covariance function ρ(x, y) is of the form ρ(|x_1-y_1|,...,|x_n - y_n|), i.e. an even function in every dimension (which surely is not always the case). But I'm not sure whether it would actually be useful to add this simplification to GaussianRandomFields even for isotropic fields. In my current project (which only deals with isotropic covariance functions so far) I'm increasing the padding in a loop until the circulant matrix is positive semidefinite (or almost, at least) or another stopping criterion is satisfied - hence usually the circulant matrix is increased and its eigenvalues are computed multiple times, and exploiting the symmetry of the first column vector seems to simplify these computations (but so far I'm still mirroring the eigenvalues of the final circulant matrix since the same "trick" can not be applied in the sampling step it seems...). However, since GaussianRandomFields computes the eigenvalues only once, I'm not sure whether exploiting the symmetry would help much. So my first comment about reducing the 2n+1 grid points in dimensions 2,...,d to 2n seems to be applicable in a more general setting and could be more useful, I guess. I'm just wondering, doesn't the current implementation of GaussianRandomFields (which does not mirror the first dimension) imply that the covariance function is of the form ρ(|x_1-y_1|,x_2 - y_2,...,x_n - y_n)? Or is there another reason for not mirroring the first dimension?

**Comment by PieterjanRobbe** [comment]
> Or is there another reason for not mirroring the first dimension?

I think my reasoning was similar as yours, i.e., because of symmetry. But then you are right that it should also be possible in the other dimensions... We might drop the support for anisotropic fields, and go with the  `brfft` or `FFTW.r2r(c, FFTW.REDFT00)` solution you proposed? I agree that the reported eigenvalues should definitely correspond to the actual eigenvalues, and not a scaled version of those.

**Comment by PieterjanRobbe** [comment]
Fixed by #12 !
